### ISSUE TITLE
rangeBehaviors are too strict

### ISSUE THREAD

**Issue by ooflorent** [facebook/relay#1208]
v0.9 introduces a way to declare `rangeBehaviors` as a function (56e52d38c6adc77bcc83926993ae39dec51bb100). This is super useful and I cannot thank you enough for this! However, they introduces a new restriction:

```js
// getRangeBehavior.js#L75
invariant(
  typeof behavior === 'string',
  'getRangeBehavior(): Expected range behavior for key `%s` to be a ' +
  'string, got `%s`.',
  call.name,
  behavior
);
```

This is an issue if your connection is defined like this:

```graphql
enum PostOrder {
  DATE_ASC
  DATE_DESC
  TITLE_ASC
  TITLE_DESC
}

type Blog implements Node {
  id: ID!
  title: String
  posts(
    first: Int
    after: ID
    order: PostOrder
    since: Int
  ): BlogPostConnection
}
```

The `invariant` test may fail on valid arguments. The following `calls` would raises an error:

```js
[
  // `order` holds an enum value.
  // Passing the value as a string is fine.
  // However, passing `null` would throws.
  {
    name: "order",
    value: null,
  },
  // `since` holds a timestamp.
  // Converting it to a string would result in an invalid query but numbers are allowed!
  {
    name: "since",
    value: 1465542134907,
  }
]
```

`rangeBehaviors` object notation isn't affected by this issue because `calls` values are serialized.

**Comment by ooflorent** [comment]
@xuorig Since you are the one who proposed this changed and implemented it. Could you bring some insight about this issue? Particularly _why `getObjectFromCalls` is denying non-string values?_

**Comment by keithpitt** [comment]
I actually hit this problem myself. I looked around the supporting code, and I couldn't find a reason as to why it's there. The error message itself is actually the same as another in the same commit https://github.com/facebook/relay/commit/56e52d38c6adc77bcc83926993ae39dec51bb100#diff-fe15bf16b6fdab1fc8ae96d4902f0f83R50 and https://github.com/facebook/relay/commit/56e52d38c6adc77bcc83926993ae39dec51bb100#diff-fe15bf16b6fdab1fc8ae96d4902f0f83R77

My current theory is that it could be a copy/paste error (since the error message itself doesn't actually make much sense in this context). Maybe it's easier if we open a PR to remove that check and continue the discussion there?

**Comment by xuorig** [comment]
Strange I don't remember adding this check in my PR @ooflorent. This was possibly added on FB side when they imported.

I agree that this really should not be there! We should either remove the check if look of `undefined` there probably...

**Comment by wincent** [comment]
> This was possibly added on FB side when they imported, so I might be missing some context.

It was added by @josephsavona with the comment "lint/flow fixes", so maybe he recalls what the lint/flow errors were that prompted the change.

**Comment by josephsavona** [comment]
I remember that I simplified the PR a bit, which prompted some type-fixing. Good catch, the type of call.value should be mixed ([here](https://github.com/facebook/relay/commit/56e52d38c6adc77bcc83926993ae39dec51bb100#diff-fe15bf16b6fdab1fc8ae96d4902f0f83R76)). Anyone interested in fixing in a PR ;-)?

**Comment by ooflorent** [comment]
@josephsavona: @xuorig has submitted a fix yesterday (#1216)
